taichi/backends/cuda/cupti_toolkit_functions.h

```diff
@@ -0,0 +1,164 @@
+#include <nvperf_host.h>
+#include <nvperf_cuda_host.h>
+#include <nvperf_target.h>
+
+// Some of the codes are copied from CUPTI/samples/extensions/
+// and modified to match Taichi's naming convensions
+
+template <typename T>
+class ScopeExit {
+ public:
+  ScopeExit(T t) : t_(t) {
+  }
+  ~ScopeExit() {
+    t_();
+  }
+  T t_;
+};
+
+template <typename T>
+ScopeExit<T> MoveScopeExit(T t) {
+  return ScopeExit<T>(t);
+};
+
+  // The fellowing macros will be #undef in ./cupti_toolkit.cpp
+
+#define CUPTI_API_CALL(api_func_call)                                      \
+  do {                                                                     \
+    CUptiResult status = api_func_call;                                    \
+    if (status != CUPTI_SUCCESS) {                                         \
+      const char *errstr;                                                  \
+      cuptiGetResultString(status, &errstr);                               \
+      fprintf(stderr, "%s:%d: error: function %s failed with error %s.\n", \
+              __FILE__, __LINE__, #api_func_call, errstr);                 \
+      exit(-1);                                                            \
+    }                                                                      \
+  } while (0)
+
+#define NVPW_API_CALL(api_func_call)                                       \
+  do {                                                                     \
+    NVPA_Status status = api_func_call;                                    \
+    if (status != NVPA_STATUS_SUCCESS) {                                   \
+      fprintf(stderr, "%s:%d: error: function %s failed with error %d.\n", \
+              __FILE__, __LINE__, #api_func_call, status);                 \
+      exit(-1);                                                            \
+    }                                                                      \
+  } while (0)
+
+#define NV_ANONYMOUS_VARIABLE_DIRECT(name, line) name##line
+
+#define NV_ANONYMOUS_VARIABLE_INDIRECT(name, line) \
+  NV_ANONYMOUS_VARIABLE_DIRECT(name, line)
+
+#define SCOPE_EXIT(func)                                      \
+  const auto NV_ANONYMOUS_VARIABLE_INDIRECT(EXIT, __LINE__) = \
+      MoveScopeExit([=]() { func; })
+
+#define RETURN_IF_NVPW_ERROR(retval, actual)                 \
+  do {                                                       \
+    NVPA_Status status = actual;                             \
+    if (NVPA_STATUS_SUCCESS != status) {                     \
+      fprintf(stderr, "FAILED: %s with error %s\n", #actual, \
+              get_nvpw_result_string(status));               \
+      return retval;                                         \
+    }                                                        \
+  } while (0)
+
+#define NVPW_STATUS_RESULT(status) \
+  case status:                     \
+    error_msg = #status;           \
+    break;
+
+static const char *get_nvpw_result_string(NVPA_Status status) {
```

Comment:
nit: Can you move these function definitions into a cpp so that we don't accidentally expose these macros or ScopeExit? (An optimal implementation of `ScopeExit` might involve some additional amount of template metaprogramming)